### PR TITLE
Preserve parens around expressions in default export declaration that start with function or class

### LINKED ISSUE #6083: Parens incorrectly dropped when trying to type a function expression
**Prettier 1.17.0**
[Playground link](https://prettier.io/playground/#N4Igxg9gdgLgprEAuEcAeAHCAnGACAEzgDMBDAVwBt8AKY8qMGAS2j0ogHMaBKPYADpQhAXzykAznhgBPDHAjE8kKBIiU4AOg6ceIADQgIGFtAnJQpbNggB3AApWE5lKQBuEZgQMgARtlIwAGs4GABlDEDmKE5kGGxyOEMACxgAW0oAdWTmeAlIsDgw51zmN1yZZHAJc0NoiThcewDONNJkMkoGwwArCTQAIQDg0LDSNLgAGWi4DtIupJA+tDDozg0ARXIIeDmFw0jsBuwq2XkJMGxmEx8MK9hMrxhk5AAOAAYDmwbMgIwqu5wY5uWaGbBwACO5GY4OapFa7SQnW6IAaaWYcQSiwka02212SPmKJgpF8jwIz2QACZDPFSMxKGsAMIQNJtKpQaCgkDkBoAFVJLmRcBEIiAA)
```sh
--parser typescript
```

**Input:**
```tsx
export default (function log() {

} as typeof console.log)
```

**Output:**
```tsx
export default function log() {} as typeof console.log;

```

**Second Output:**
```tsx
SyntaxError: ';' expected. (1:37)
> 1 | export default function log() {} as typeof console.log;
    |                                     ^
  2 | 
```
**Expected behavior:**

Parens should not be dropped as they are required in this case.  Workaround is to use an arrow function:

```tsx
export default const log: typeof console.log = () => {}
```

### PR DESCRIPTION
**Pull request by duailibe**
Closes #6083

**✨[Try the playground for this PR](https://prettier.io/playground-redirect)✨**

### CODE REVIEW

**Comment by thorn0** [comment]
It's not something specific to as-expressions. The same happens with other default-exported binary expressions if their left argument is a function expression, function declaration, or an arrow function. And not only in TypeScript. Babel doesn't parse this too:

**Prettier 1.17.1**
[Playground link](https://prettier.io/playground/#N4Igxg9gdgLgprEAuEcAeAHCAnGACAEzgDMBDAVwBt9jyowYBLaPSiAcwAoBKYAXzwAqPACYA3AB0oIADQgIGJtADOyUKWzYIAdwAKGhKpSkAbhEYFZIAEbZSYANZwYAZQz3GUdshjZycOQALGABbSgB1QMZ4ZXcwOBdDaMYTaIBPZHBlVTlPZThcXTt2ENJkMkp8uQArZTQAITtHZxdSELgAGU84ctJKgJBatBdPdko4AEVyCHhe-rl3bHzsTJg0jDhlMGxGRSsMHdhwixhA5AAOAAYFrXzwuwxMg82Ckx65AEdp+CKFIxBSMoALRQOBwIiWOTYOBfRjQoqkEplJAVKogfIhRg+PwDZSjcZTGY9FF9NEwUjWY4EU7IERyXykRiUUYAYQgIVKmSg0HeIHI+QAKhSjKi4Hw+EA)
```sh
--parser typescript
```

**Input:**
```tsx
export default function log(){} * 2;

```

**Output:**
```tsx
SyntaxError: Expression expected. (1:33)
> 1 | export default function log(){} * 2;
    |                                 ^
  2 | 
```

**Comment by duailibe** [comment]
@thorn0 Can you think of practical cases that aren't yet covered?

The one you mentioned is correct per the spec, but doesn't happen in practice because isn't valid in the runtime, so we shouldn't need to worry about that.

**Comment by thorn0** [comment]
It surely feels like a stretch, but still, I can imagine somebody doing things like these:
```js
export default (function(){  /* ... */ }.toString());
export default (class WeirdSingleton { /* ... */ }.getInstance());
```
And don't forget about the upcoming pipeline operator. It totally makes sense for a function or a class expression to be its left-hand operand.

**Comment by duailibe** [comment]
Thanks @thorn0, those are more practical use cases.

Unfortunately because of the format of the AST it's a bit hard to detect when there's a `function` or `class` keyword in the start of the expression.

We do have some code to try and guess what is the "left most" node in an expression to add ASI protection. I'll see if we can reuse some of the code for that. Otherwise, we have to manually handle all possible cases.

**Comment by duailibe** [comment]
@lydell @thorn0 What do you think of this solution now?

**Comment by thorn0** [comment]
Tagged template literals get extra parens:

**Prettier pr-6133**
[Playground link](https://deploy-preview-6133--prettier.netlify.com/playground/#N4Igxg9gdgLgprEAuEcAeAHCAnGACAEzgDMBDAVwBt8AKY8qMGAS2hoEo9gBfdgA1J8A3CAA0ICBhbQAzslCls2CAHcACooRyUpAG4RmBMSABG2UmADWcGAGUMF5lADmyGNnJxxACxgBbSgB1b2Z4GQcwOFstUOZdUIBPZHAZOXEnGThcNXNnP1JkMkpM8QArGTQAIXMrG1tSPzgAGSc4QtJirxBytFsnZ0o4AEVyCHh2zvEHbEzsZJNSEwTKaGMMbCcYQMMYb2QADgAGKeVMwPMMZPW4Wd028QBHUfgcyW0QUhkAWig4OCIjOJsHAnsxgTlSHkCkgiiUQJk-Mw3B4ujJ+oMRmM2jCOnCYIttgRdsgAEzidykZiUfoAYQgfnyySg0HuIHImQAKottLC4NxuEA)
```sh
--parser babylon
```

**Input:**
```jsx
export default (function() {})`a`;
```

**Output:**
```jsx
export default ((function() {})`a`);

```

File: src/language-js/needs-parens.js [hunk 1]
```diff
@@ -155,6 +159,23 @@ function needsParens(path, options) {
     return true;
   }
 
+  // `export default function` or `export default class` can't be followed by
+  // anything after. So an expression like `export default (function(){}).toString()`
+  // needs to be followed by a parentheses
+  if (parent.type === "ExportDefaultDeclaration") {
+    // IIFE already include parenthesis between the expression and invocation
+    // so we don't need to handle here.
```
Comment (thorn0) [inline src/language-js/needs-parens.js]: Would be good to have a test with IIFE.

Comment (duailibe) [reply src/language-js/needs-parens.js]: I was already adding those 👍 
Thanks for the suggestion

**Comment by duailibe** [comment]
@thorn0 Thank you for testing!

Now it only wraps the expression being exported if the `function` or `class` isn't already wrapped.

File: tests/export_default/__snapshots__/jsfmt.spec.js.snap [hunk 2]
```diff
@@ -13,3 +13,59 @@ export default (class {}[1] = 1);
...
+
+=====================================output=====================================
+export default (class {}.getInstance());
+
+================================================================================
+`;
+
+exports[`function_in_template.js 1`] = `
+====================================options=====================================
+parsers: ["flow", "typescript"]
```
Comment (thorn0) [inline tests/export_default/__snapshots__/jsfmt.spec.js.snap]: I wonder why these tests aren't run using `babel`.

File: src/language-js/needs-parens.js [hunk 3]
```diff
@@ -839,4 +851,31 @@ function isFollowedByRightBracket(path) {
...
+      // in this case we don't need to add extra parens
+      !needsParens(path, options)
+    );
+  }
+
+  if (!hasNakedLeftSide(node)) {
+    return false;
+  }
+
+  return path.call.apply(
```
Comment (thorn0) [inline src/language-js/needs-parens.js]: Why not use spread syntax here instead of `apply` and `concat`?

**Comment by thorn0** [comment]
Overall looks great, but I did manage to produce superfluous parens: 

**Prettier pr-6133**
[Playground link](https://deploy-preview-6133--prettier.netlify.com/playground/#N4Igxg9gdgLgprEAuEcAeAHCAnGACAEzgDMBDAVwBt8AKY8qMGAS2hoEpgBfPAHwD48pdgAMRAbhAAaEBAwtoAZ2ShS2bBADuABTUJlKUgDcIzAtJAAjbKTABrODADKGW8ygBzZDGzk4MgAsYAFtKAHUA5nhFVzA4J30o5iMogE9kcEVlGXdFOFxtGw9g0mQySjyZACtFNAAhG3tHJ1JguAAZdzgy0gr-EBq0J3cPSjgARXIIeB6+mVdsPOwMy1JLOEoLDGx3GDCzGADkAA4ABnmNPLCbDAztuCWjbpkARyn4QrkDEFJFAFooHA4ERzDJsHA3sxwYVSMVSkhypUQHlgsxvL5+ooRmNJtNugjekiYGt9gRDsgAEwyHykZiUEYAYQgwRKGSg0GeIHIeQAKmsDIi4FwuEA)
```sh
--parser babel
```

**Input:**
```jsx
export default (function(){} |> a)``;
```

**Output:**
```jsx
export default ((function() {} |> a)``);
```

**Comment by duailibe** [comment]
@thorn0 Thanks for beta testing!

After years of Prettier this is the first bug report we've had, I think we're on the safe side now. From a practical point of view, we've covered enough cases IMO.

**Pull request merged**